engine/consensus/compliance/engine.go

```diff
@@ -66,49 +53,100 @@ func NewEngine(
 	me module.Local,
 	prov network.Engine,
 	core *Core) (*Engine, error) {
-	e := &Engine{
-		unit:             engine.NewUnit(),
-		log:              log.With().Str("compliance", "engine").Logger(),
-		me:               me,
-		mempool:          core.mempool,
-		metrics:          core.metrics,
-		headers:          core.headers,
-		payloads:         core.payloads,
-		state:            core.state,
-		tracer:           core.tracer,
-		prov:             prov,
-		core:             core,
-		pendingEventSink: make(EventSink),
-		blockSink:        make(EventSink),
-		voteSink:         make(EventSink),
-	}
-
-	var err error
-	// register the core with the network layer and store the conduit
-	e.con, err = net.Register(engine.ConsensusCommittee, e)
-	if err != nil {
-		return nil, fmt.Errorf("could not register core: %w", err)
-	}
 
 	// FIFO queue for block proposals
-	e.pendingBlocks, err = fifoqueue.NewFifoQueue(
+	pendingBlocks, err := fifoqueue.NewFifoQueue(
 		fifoqueue.WithCapacity(defaultBlockQueueCapacity),
-		fifoqueue.WithLengthObserver(func(len int) { e.mempool.MempoolEntries(metrics.ResourceBlockProposalQueue, uint(len)) }),
+		fifoqueue.WithLengthObserver(func(len int) { core.mempool.MempoolEntries(metrics.ResourceBlockProposalQueue, uint(len)) }),
 	)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create queue for inbound receipts: %w", err)
 	}
 
 	// FIFO queue for block votes
-	e.pendingVotes, err = fifoqueue.NewFifoQueue(
+	pendingVotes, err := fifoqueue.NewFifoQueue(
 		fifoqueue.WithCapacity(defaultVoteQueueCapacity),
-		fifoqueue.WithLengthObserver(func(len int) { e.mempool.MempoolEntries(metrics.ResourceBlockVoteQueue, uint(len)) }),
+		fifoqueue.WithLengthObserver(func(len int) { core.mempool.MempoolEntries(metrics.ResourceBlockVoteQueue, uint(len)) }),
 	)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create queue for inbound approvals: %w", err)
 	}
 
-	return e, nil
+	// define message queueing behaviour
+	handler := engine.NewMessageHandler(
+		log,
+		engine.Pattern{
+			Match: func(msg *engine.Message) bool {
+				_, ok := msg.Payload.(*messages.BlockProposal)
+				return ok
+			},
+			BeforeStore: []engine.OnMessageFunc{
+				func(_ *engine.Message) {
+					core.metrics.MessageReceived(metrics.EngineCompliance, metrics.MessageBlockProposal)
+				},
+			},
+			Store: pendingBlocks,
+		},
+		engine.Pattern{
+			Match: func(msg *engine.Message) bool {
+				_, ok := msg.Payload.(*events.SyncedBlock)
+				return ok
+			},
+			Map: func(msg *engine.Message) *engine.Message {
+				syncedBlock := msg.Payload.(*events.SyncedBlock)
+				return &engine.Message{
+					OriginID: msg.OriginID,
+					Payload: &messages.BlockProposal{
+						Payload: syncedBlock.Block.Payload,
+						Header:  syncedBlock.Block.Header,
+					},
+				}
+			},
+			BeforeStore: []engine.OnMessageFunc{
+				func(_ *engine.Message) {
+					core.metrics.MessageReceived(metrics.EngineCompliance, metrics.MessageSyncedBlock)
+				},
+			},
```

Comment:
> @durkmurder : Also can we have a way to filter out events after matching but before storing into queue?
For instance in sealing engine we are providing metrics but ignore the message completely, check here

We can track the metrics when consuming the message from the queue. And then we could even get rid of this `BeforeStore` to keep it simple. 

The use case of `BeforeStore` here is just for metrics. And the reason it's a slice is so that by default the empty slice means nothing to do before storing in into the queue.

We could track the metrics in `processAvailableMessages` instead of here. 

If we would like to save the metrics before storing, we actually have the network received message metrics already, it would be a duplication